openstudiocore/src/model/BuildingUnit.cpp

```diff
@@ -0,0 +1,430 @@
+#include "RenderingColor.hpp"
+#include "RenderingColor_Impl.hpp"
+#include "ModelExtensibleGroup.hpp"
+
+#include <utilities/idd/OS_BuildingUnit_FieldEnums.hxx>
+#include <utilities/idd/IddEnums.hxx>
+#include <utilities/idd/IddFactory.hxx>
+
+namespace openstudio {
+namespace model {
+namespace detail {
+
+  BuildingUnit_Impl::BuildingUnit_Impl(const IdfObject &idfObject, Model_Impl *model, bool keepHandle)
+    : ModelObject_Impl(idfObject, model, keepHandle)
+  {
+    OS_ASSERT(idfObject.iddObject().type() == BuildingUnit::iddObjectType());
+  }
+
+  BuildingUnit_Impl::BuildingUnit_Impl(const openstudio::detail::WorkspaceObject_Impl &other,
+                                       Model_Impl *model,
+                                       bool keepHandle)
+    : ModelObject_Impl(other, model, keepHandle)
+  {
+    OS_ASSERT(other.iddObject().type() == BuildingUnit::iddObjectType());
+  }
+
+  BuildingUnit_Impl::BuildingUnit_Impl(const BuildingUnit_Impl &other,
+                                       Model_Impl *model,
+                                       bool keepHandle)
+    : ModelObject_Impl(other, model, keepHandle)
+  {}
+
+  const std::vector<std::string>& BuildingUnit_Impl::outputVariableNames() const
+  {
+    static std::vector<std::string> result;
+    if (result.empty()){
+    }
+    return result;
+  }
+
+  IddObjectType BuildingUnit_Impl::iddObjectType() const {
+    return BuildingUnit::iddObjectType();
+  }
+
+  boost::optional<RenderingColor> BuildingUnit_Impl::renderingColor() const
+  {
+    return getObject<ModelObject>().getModelObjectTarget<RenderingColor>(OS_BuildingUnitFields::RenderingColor);
+  }
+
+  bool BuildingUnit_Impl::setRenderingColor(const RenderingColor& renderingColor)
+  {
+    return setPointer(OS_BuildingUnitFields::RenderingColor, renderingColor.handle());
+  }
+
+  void BuildingUnit_Impl::resetRenderingColor()
+  {
+    bool ok = setString(OS_BuildingUnitFields::RenderingColor, "");
+    OS_ASSERT(ok);
+  }
+
+  boost::optional<std::string> BuildingUnit_Impl::buildingUnitType() const
+  {
+    return getString(OS_BuildingUnitFields::BuildingUnitType, true, true);
+  }
+
+  bool BuildingUnit_Impl::setBuildingUnitType(const std::string& buildingUnitType)
+  {
+    bool ok = setString(OS_BuildingUnitFields::BuildingUnitType, buildingUnitType);
+    return ok;
+  }
+
+  void BuildingUnit_Impl::resetBuildingUnitType()
+  {
+    bool ok = setString(OS_BuildingUnitFields::BuildingUnitType, "");
+    OS_ASSERT(ok);
+  }
+
+  std::vector<Space> BuildingUnit_Impl::spaces() const
+  {
+    return getObject<ModelObject>().getModelObjectSources<Space>(Space::iddObjectType());
+  }
+
+  std::vector<std::string> BuildingUnit_Impl::featureNames() const
+  {
+    std::vector<std::string> names;
+    for (const ModelExtensibleGroup& group : castVector<ModelExtensibleGroup>(extensibleGroups())) {
+      boost::optional<std::string> name = group.getString(OS_BuildingUnitExtensibleFields::BuildingUnitFeatureName);
```

Comment:
So, should I do some sort of check for duplicate names here and log an error if there are duplicates?